### DISCUSSION TITLE
cpp/memory/shared_ptr

### DISCUSSION THREAD

**Discussion by giscus** [imarvinle/CSGuide#29] [Announcements]
# cpp/memory/shared_ptr

编程指北: 计算机系统学习指南, 操作系统, 计算机网络, C++, Java, 算法, 数据结构

https://csguide.cn/cpp/memory/shared_ptr.html

**Comment by coolhuhu** [comment]
作者你好，你在该篇文章中的”线程安全性“部分写道”其实 shared_ptr 线程不安全主要来自于引用计数有并发更新的风险，当然引用计数本身也可以使用原子atomic”，“引用计数本身也可以使用原子atomic”的表述是不是不太严谨？shared_ptr的引用计数本身是原子性的，在 https://en.cppreference.com/w/cpp/memory/shared_ptr#Implementation_notes 链接中有说明，如下是原文：

To satisfy thread safety requirements, the reference counters are typically incremented using an equivalent of std::atomic::fetch_add with std::memory_order_relaxed (decrementing requires stronger ordering to safely destroy the control block).

当然也许是我理解有偏差。

**Comment by MichaelAbel1** [comment]
手写shared_ptr中的赋值操作符操作中的release()是否没必要？

**Reply by MichaelAbel1** [reply]
噢 变换指针的时候有必要

**Comment by valuejhxu** [comment]
智能指针的实现中，引用计数`count_`是私有成员变量，不能通过`.`来获取

**Reply by valuejhxu** [reply]
好像可以
